### ISSUE TITLE
[Feature] passing `this` to mutators

### ISSUE THREAD

**Issue by SzNagyMisu** [vuex-orm/vuex-orm#441]
Would it be possible to pass the whole model instance to the [mutators](https://vuex-orm.github.io/vuex-orm/guide/model/accessors-and-mutators.html#defining-mutators)?

Sometimes I need to update a property when another property changes.
For example, in our app we store amounts of money along with the currency and the exchange rate for converting to HUF. When a user sets the currency to `"HUF"`, I want the exchange rate to change automatically to `1.0`.
For now, it is solved by a [watcher](https://vuejs.org/v2/guide/computed.html#Watchers) in the component, but I feel like it is rather model logic.
I tried to solve it via mutators, but they don't have access to values of other properties. On the other hand it may not be a good idea to give them access, as their only task is to mutate the data passed to the setter...

Any ideas?

**Comment by kiaking** [comment]
@SzNagyMisu Interesting idea! In your case, I think you could use simple getter to achieve what you want.

```js
class Price extends Model {
  static entity = 'prices'

  static fields () {
    return {
      id: this.attr(null),
      currency: this.string('NOT_HUF'),
      price: this.number(0)
    }
  }

  get value () {
    if (this.currency === 'HUF') {
      // Do the HUF thing.
    }

    // Else, do something diferent.
  }
}
```

However, I think to be able to pass the whole model to mutator is nice to have. I'll mark this issue as an enhancement issue!

**Comment by SzNagyMisu** [comment]
@kiaking Yes, I have considered that possibility. And I have two reasons not to use it:

1. When passing data to the backend API the model instances go through JSON stringification. That is done [with the use of the method `toJSON()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#Description) (defined [here](https://github.com/vuex-orm/vuex-orm/blob/60d38046c910f5227353c1aada8a9f4f25440a4d/src/model/Model.ts#L818) by `vuex-orm`). Since it does not include getters (which is good, I think), I would be forced to override the method `toJSON()` every time I need to pass such a computed value to the backend. This feels a bit **hackish** for me...
2. Anyway, I **don't** want to **store false data** in my models. If the value of `currency` is `'HUF'`, I would rather have the value of `exchangeRate` set to `1.0` then leave it something else and define a getter to have the correct value...

So, for now, I will stay with the component warchers.

But thank you for labelling this issue as enhancement! And tell me if I can do anything for the cause :smiley:

**Comment by eldomagan** [comment]
@SzNagyMisu You can achieve this using beforeUpdate lifecycle hook. I have done the same thing on a project I'm working on.

```js
class Price extends Model {
  static entity = 'prices'

  static beforeUpdate (record) {
    if (record.currency === 'HUF') {
      record.exchangeRate = 1
    }
  }
}
```

**Comment by SzNagyMisu** [comment]
Great idea, @eldomagan ! It may partly solve my problem.

The only thing is that I need to change most of the values on-the-fly. So, for example, if I have a form with a field for `currency` and another for `exchangeRate`, changing `currency` to `'HUF'` must change the value of `exchangeRate` to `1.0`, **immediately**. And there is no lifecycle hook for that as yet...

**Comment by SzNagyMisu** [comment]
@kiaking Can I provide a PR for this?

**Comment by kiaking** [comment]
@SzNagyMisu Yes sure! That would be awesome!

**Comment by SzNagyMisu** [comment]
OK, I just started the PR when I realized, I was completely wrong!

Somehow I imagined it worked like mutating the value when setting the property. So I thought `record.field = value` would trigger the the mutators. But it does not. Now, I understand better what @eldomagan  said.

So because of the following reasons

1. the mutators cannot possibly know about the instance
2. it would not help my case with changes in `currency` triggering changes in `exchangeRate` on-the-fly

I decided to close the issue.
